core/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -663,6 +673,37 @@ public boolean canDeleteIndexContents(Index index, IndexSettings indexSettings)
     }
 
     /**
+     * Verify that the contents on disk for the given index is deleted; if not, delete the contents.
+     * This method assumes that an index is already deleted in the cluster state and/or explicitly
+     * through index tombstones.
+     * @param index {@code Index} to make sure its deleted from disk
+     * @param clusterState {@code ClusterState} to ensure the index is not part of it
+     * @return IndexMetaData for the index loaded from disk
+     */
+    @Nullable
+    public IndexMetaData verifyIndexIsDeleted(final Index index, final ClusterState clusterState) {
+        // this method should only be called when we know the index is not part of the cluster state
+        assert clusterState.metaData().hasIndex(index.getName()) == false;
+        if (nodeEnv.hasNodeFile() && FileSystemUtils.exists(nodeEnv.indexPaths(index))) {
+            final IndexMetaData metaData;
+            try {
+                metaData = metaStateService.loadIndexState(index);
+            } catch (IOException e) {
+                logger.warn("[{}] failed to load state file from a stale deleted index, folders will be left on disk", e, index);
+                return null;
+            }
+            final IndexSettings indexSettings = buildIndexSettings(metaData);
+            try {
+                deleteIndexStoreWithPredicate("stale deleted index", index, indexSettings, ALWAYS_TRUE);
+            } catch (IOException e) {
+                logger.warn("[{}] failed to delete index on disk", e, metaData.getIndex());
```

Comment:
The `deleteIndexStoreWithPredicate` method adds the deletion to the pending queue for trying again.